include/mixer.h

```diff
@@ -151,6 +160,8 @@ class MixerChannel {
 	void AddSamples_s16u(uint16_t len, const uint16_t *data);
 	void AddSamples_m32(uint16_t len, const int32_t *data);
 	void AddSamples_s32(uint16_t len, const int32_t *data);
+	void AddSamples_mfloat(uint16_t len, const float *data);
+	void AddSamples_sfloat(uint16_t len, const float *data);
```

Comment:
Very nice! 

Quite a few audio devices use float calculations at some point in their sample calculations, which get rounded or cast back to along the way - but now we can "stay native" for the entire pipeline.

Even unexpected places, like decoding compressed CD audio tracks in Vorbis, for example:

It would prefer to give us access to it's internal float arrays:

``` c++
int stb_vorbis_get_frame_float(stb_vorbis *f, int *channels, float ***output)
{
   int len, right,left,i;
   if (IS_PUSH_MODE(f)) return error(f, VORBIS_invalid_api_mixing);

   if (!vorbis_decode_packet(f, &len, &left, &right)) {
      f->channel_buffer_start = f->channel_buffer_end = 0;
      return 0;
   }

   len = vorbis_finish_frame(f, len, left, right);
   for (i=0; i < f->channels; ++i)
      f->outputs[i] = f->channel_buffers[i] + left;

   f->channel_buffer_start = left;
   f->channel_buffer_end   = left+len;

   if (channels) *channels = f->channels;
   if (output)   *output = f->outputs;
   return len;
}
```

But previously we decoded using their 16-bit accessor:

``` c++
int stb_vorbis_get_frame_short_interleaved(stb_vorbis *f, int num_c, short *buffer, int num_shorts)
{
   float **output;
   int len;
   if (num_c == 1) return stb_vorbis_get_frame_short(f,num_c,&buffer, num_shorts);
   len = stb_vorbis_get_frame_float(f, NULL, &output);
   if (len) {
      if (len*num_c > num_shorts) len = num_shorts / num_c;
      convert_channels_short_interleaved(num_c, buffer, f->channels, output, 0, len);
   }
   return len;
}
```

Which first gets the samples as float.. and then coverts, look at all those `for` loops (pains me!):

``` c++
static void convert_channels_short_interleaved(int buf_c, short *buffer, int data_c, float **data, int d_offset, int len)
{
   int i;
   check_endianness();
   if (buf_c != data_c && buf_c <= 2 && data_c <= 6) {
      assert(buf_c == 2);
      for (i=0; i < buf_c; ++i)
         compute_stereo_samples(buffer, data_c, data, d_offset, len);
   } else {
      int limit = buf_c < data_c ? buf_c : data_c;
      int j;
      for (j=0; j < len; ++j) {
         for (i=0; i < limit; ++i) {
            FASTDEF(temp);
            float f = data[i][d_offset+j];
            int v = FAST_SCALED_FLOAT_TO_INT(temp, f,15);//data[i][d_offset+j],15);
            if ((unsigned int) (v + 32768) > 65535)
               v = v < 0 ? -32768 : 32767;
            *buffer++ = v;
         }
         for (   ; i < buf_c; ++i)
            *buffer++ = 0;
      }
   }
}
```


These little improvement

Reply:
Cheers, and +1 for pointing out the Vorbis decoding path! I've created a cleanup ticket for this, please feel free to add more cases to it where such improvements could be made.

https://github.com/dosbox-staging/dosbox-staging/issues/1720

Reply:
Thanks for opening that ticket! Added a couple.